File: sort.go
package algo

// bubbleSort O(N^2)
func bubbleSort[T numbers](arr []T) []T {
	sorted := false

	for !sorted {
		sorted = true

		for i := 0; i < len(arr)-1; i++ {
			if arr[i] > arr[i+1] {
				arr[i], arr[i+1] = arr[i+1], arr[i]
				sorted = false
			}
		}
	}

	return arr
}

// selectionSort O(N^2)
func selectionSort[T numbers](arr []T) []T {
	for i := 0; i < len(arr)-1; i++ {
		lowestNumIdx := i
		for j := i + 1; j < len(arr); j++ {
			if arr[j] < arr[lowestNumIdx] {
				lowestNumIdx = j
			}
		}

		if lowestNumIdx != i {
			temp := arr[i]
			arr[i] = arr[lowestNumIdx]
			arr[lowestNumIdx] = temp
		}
	}

	return arr
}

func insertionSort[T numbers](arr []T) []T {
	for i := 1; i < len(arr); i++ {
		tempVal := arr[i]
		position := i - 1

		for position >= 0 {
			if arr[position] > tempVal {
				arr[position+1] = arr[position]
				position--
			} else {
				break
			}
		}

		arr[position+1] = tempVal
	}

	return arr
}

File: sort.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();
const expect = std.testing.expect;

fn bubbleSort(comptime T: type, array: []T) !void {
    var sorted = false;

    while (!sorted) {
        sorted = true;

        var i: u8 = 0;
        while (i < array.len - 1) : (i += 1) {
            if (array[i] > array[i + 1]) {
                const temp = array[i];
                array[i] = array[i + 1];
                array[i + 1] = temp;

                sorted = false;
            }
        }
    }
}

fn selectionSort(comptime T: type, array: []T) !void {
    var i: usize = 0;
    while (i < array.len - 1) : (i += 1) {
        var lowestNumIdx: usize = i;

        var j = i + 1;
        while (j < array.len) : (j += 1) {
            if (array[j] < array[lowestNumIdx])
                lowestNumIdx = j;
        }

        if (lowestNumIdx != i) {
            const temp = array[i];
            array[i] = array[lowestNumIdx];
            array[lowestNumIdx] = temp;
        }
    }
}

fn insertionSort(comptime T: type, array: []T) !void {
    var i: usize = 1;
    while (i < array.len) : (i += 1) {
        const tmp = array[i];
        var j = i;

        while (j > 0 and array[j - 1] > tmp) : (j -= 1) {
            array[j] = array[j - 1];
        }
        array[j] = tmp;
    }
}

// tests

test "bubble sort" {
    const exp = [_]u8{ 10, 15, 25, 35, 45, 55, 65 };
    var input = [_]u8{ 65, 55, 45, 35, 25, 15, 10 };

    try bubbleSort(u8, input[0..]);

    // try stdout.print("{any}\n", .{input});

    try expect(std.mem.eql(u8, exp[0..], input[0..]));
}

test "selection sort" {
    const exp = [_]u8{ 10, 15, 25, 35, 45, 55, 65 };
    var input = [_]u8{ 65, 55, 45, 35, 25, 15, 10 };

    try selectionSort(u8, input[0..]);

    // try stdout.print("{any}\n", .{input});

    try expect(std.mem.eql(u8, exp[0..], input[0..]));
}
test "insertion sort" {
    const exp = [_]u8{ 10, 15, 25, 35, 45, 55, 65 };
    var input = [_]u8{ 65, 55, 45, 35, 25, 15, 10 };

    try insertionSort(u8, input[0..]);

    // try stdout.print("{any}\n", .{input});

    try expect(std.mem.eql(u8, exp[0..], input[0..]));
}
